File: Editorial/16_3Sum_Closest/CSharp/1_Two_Pointers.cs
public class Solution {
    public int ThreeSumClosest(int[] nums, int target) {
        int diff = Int32.MaxValue;
        int sz = nums.Length;
        Array.Sort(nums);
        for (int i = 0; i < sz && diff != 0; ++i) {
            int lo = i + 1;
            int hi = sz - 1;
            while (lo < hi) {
                int sum = nums[i] + nums[lo] + nums[hi];
                if (Math.Abs(target - sum) < Math.Abs(diff)) {
                    diff = target - sum;
                }

                if (sum < target) {
                    ++lo;
                } else {
                    --hi;
                }
            }
        }

        return target - diff;
    }
}

File: Editorial/16_3Sum_Closest/Go/1_Two_Pointers.go
func threeSumClosest(nums []int, target int) int {
    diff := math.MaxInt32
    sort.Ints(nums)
    for i := 0; i < len(nums) && diff != 0; i++ {
        lo := i + 1
        hi := len(nums) - 1
        for lo < hi {
            sum := nums[i] + nums[lo] + nums[hi]
            if abs(target-sum) < abs(diff) {
                diff = target - sum
            }
            if sum < target {
                lo++
            } else {
                hi--
            }
        }
    }
    return target - diff
}

func abs(x int) int {
    if x < 0 {
        return -x
    }
    return x
}
